### DISCUSSION TITLE
Python API to map between md5 and a git tag

### DISCUSSION THREAD

**Discussion by mohitanchlia** [treeverse/dvc#9964] [General]
It would be good to have an API to map from MD5 to a git tag

It would also be nice to know the tag that was used when passing a null in "rev" in python API

**Comment by efiop** [comment]
Converted to discussion since this is not actionable right now.

We don't really have that mapping internally, so this is more of an active operation that every user can do himself by parsing his own git repo. I'm not sure we have a use for it ourselves that will justify us implementing it as a feature.

**Comment by mohitanchlia** [comment]
Use case for us is to be able to get the tag name and the revision/version using Python API and log this info to mlflow for a full view of the experiment and tracking. Is there a utility function that is available in the meantime?

**Reply by shcheklein** [reply]
> It would also be nice to know the tag that was used when passing a null in "rev" in python API

could you please clarify, what API are we talking about? 

> map from MD5 to a git tag

where / how do you get the MD5 itself? could you give some examples of the workflow, some code snippets so that we understand better the use case?

thanks!

**Reply by mohitanchlia** [reply]
For instance we use this API to get the url:

`def get_url(path: str,
            repo: str = None,
            rev: str = None,
            remote: str = None) -> str`

Rev is optional and in this case it just uses the most recent branch. We call this API during running an experiment, as part of logging all the information for reproducibility we also want to get handle on the git info/tag and the md5 that DVC ended up using internally. This would help us know exactly which version was used and the git info. This is our requirement that we currently don't have a good way other than hard coding both the tag and the version to be able to persist it in mlflow along with the experiment info.

**Reply by shcheklein** [reply]
Are you using `dvc exp` to run things?

Are you trying to get git tags also from the latest commit (HEAD)?

Are using this specifically on the data tracked by DVC (`.dvc` files)?

**Reply by mohitanchlia** [reply]
data tracked by .dvc files that are checked in git. We need to persist the associated git tag and md5 used to pull the files for

**Reply by shcheklein** [reply]
so, what happens if you have a `.dvc` file in the workspace that is not committed into Git yet (so there are also no tags yet for this new version)? or do you expect people always doing `git commit` or users don't change `.dvc` often?

To get git tags I would recommend using on the existing Git wrappers for Python - https://github.com/libgit2/pygit2 or https://github.com/jelmer/dulwich or just even run `git describe` directly in Python.

Would that work for your use case?

Btw, out of curiosity - have you seen Studio and have you tried `dvc exp` - it could potentially solve some of this issues out of the box, I would love to hear more thoughts / feedback on that.

**Reply by mohitanchlia** [reply]
What we were hoping was to have DVC give us all the details instead of relying on a programmer or data scientists to execute git wrapper commands or something else. Both md5 and git information are available to DVC, it would be extremely helpful for us to get a tight integration of DVC to our experiment info in MLFlow.

We are currently using mlflow and that's why we haven't looked at dvc exp. Not sure what support we would get to make that switch at this point.

**Reply by shcheklein** [reply]
It's a reasonable request. The challenge here is that Git provides tremendous amount of information. DVC provides additional information on top of that (`md5`). I personally don't think we could meaningfully replace Git Python libraries (that already do their job) well to extract all the possible Git information.

Do you have in mind in how that interface would look like?

> instead of relying on a programmer or data scientists to execute git wrapper commands or something else.

absolutely. It should be packaged a single wrapper that is being used by DS / ML team. They definitely should not have to deal with Git or even DVC `ger_url` to my mind. Usually it can be done once, standardized, and even made as part of some callbacks (depends on what framework you use).

> Not sure what support we would get to make that switch at this point.

not 100% sure I understand the support part here, but nw, it's absolutely up to you. I was asking only to get some feedback in case you tried it before.

**Reply by mohitanchlia** [reply]
Currently, we don't have a good way to get a handle of the location of the dataset that we want to run in an experiment other than calling something like get_url. get_url somehow is able to determine the correct version of the file for a given git tag or gives latest when version is null. I was wondering if we could use the same logic and return additional metadata eg: get_metadata(revision: optional) where revision is a git tag. When None is specified it returns the latest information. Metadata struct could be {git_info: {}, file_version:md5, ...}

By "support" I meant support within the organization. We are currently committed to using mlflow. I believe mlflow and dvc exp are competing technologies.

**Reply by mohitanchlia** [reply]
@shcheklein Following up to see if you need any more info

**Reply by shcheklein** [reply]
@mohitanchlia hey, no, I think the use case is clear and your need is clear. We can't prioritize building this API on our end atm. It's still not clear to me if we want to cover the Git portion of metadata via DVC API. I would first recommend still to try to use a combination of DVC + Git Python library to get this combination. If you do this please share (or ask questions) and we can publish this as an example in docs as a first step I think.
